src/dune_engine/build_system.ml

```diff
@@ -432,7 +432,13 @@ end = struct
            without sandboxing. We just need to make sure we clean up all stale
            directory targets before running the rule and then we can discover
            all created files right in the build directory. *)
-        if not (Path.Build.Set.is_empty targets.dirs) then
+        if
+          (not (Path.Build.Set.is_empty targets.dirs))
+          &&
+          match Action.is_useful_to_sandbox action with
```

Comment:
This change (and the change below) looks very odd. If `Action.is_useful_to_sandbox action = Clearly_not`, why is `sandbox` not set to `None`?

Reply:
What do you mean by the sandbox being set to `None`? As far as I can tell, I'm just changing a validation check here.

Perhaps I should explain my intention. Before this change, actions outside of a sandbox couldn't create directory targets even if `Action.is_useful_to_sandbox action = Clearly_not` for them. My change relaxes the check to allow unsandboxed actions to create directory targets if sandboxing is a no-op for them.

Reply:
Sorry, I still don't follow. You say you relax the check, but how is going from `x` to `x && y` relaxing? Isn't this strengthening the check?

This PR is it has no description or clear motivation (apart from the title that doesn't tell me much since I've no idea what "installation" is), which makes reviewing the changes pretty hard.

Reply:
> Sorry, I still don't follow. You say you relax the check, but how is going from x to x && y relaxing? Isn't this strengthening the check?

It is relaxing in the sense that it *allows* actions which produce a directory and are unaffected by sandboxing to do so *without* us specifying the action is sandboxed.

> This PR is it has no description or clear motivation (apart from the title that doesn't tell me much since I've no idea what "installation" is), which makes reviewing the changes pretty hard.

Note that there's a linked issue that you commented at before. However, I updated the description to include the relevant bits here as well.